pretty_midi/instrument.py

```diff
@@ -113,6 +120,29 @@ def get_piano_roll(self, fs=100, times=None):
             piano_roll[note.pitch,
                        int(note.start*fs):int(note.end*fs)] += note.velocity
 
+        # Process sustain pedals
+        if pedal_threshold is not None:
+            CC_SUSTAIN_PEDAL = 64
+            time_pedal_on = 0
+            is_pedal_on = False
+            for cc in [_e for _e in self.control_changes
+                       if _e.number == CC_SUSTAIN_PEDAL]:
+                time_now = int(cc.time*fs)
+                is_current_pedal_on = (cc.value > pedal_threshold)
+                if not is_pedal_on and is_current_pedal_on:
+                    time_pedal_on = time_now
+                    is_pedal_on = True
+                elif is_pedal_on and not is_current_pedal_on:
+                    # For each pitch, a sustain pedal "retains"
+                    # the maximum velocity up to now due to
+                    # logarithmic nature of human loudness perception
+                    subpr = piano_roll[:, time_pedal_on:time_now]
+
+                    # Take the running maximum
+                    pedaled = np.maximum.accumulate(subpr, axis=1)
```

Comment:
I'm not totally sure this is the correct way to handle repeated notes with sustain. The way we did it on magenta was to extend each note within a sustain to either the beginning of the next note of the same pitch or the end of the sustain period, whichever happens first:

https://github.com/tensorflow/magenta/blob/master/magenta/music/sequences_lib.py#L972

The pianoroll equivalent of that would be to just retain whatever the most recent velocity was, not the maximum. My argument for using the most recent velocity would be that if you play a note with high velocity on a piano with sustain on, then wait a long time, then play the same note with low velocity, you won't hear a loud note, you'll hear a soft note.

That said, I could see it going the other way too (play a soft note immediately after a loud note), but I think retaining the most recent velocity better captures what's going on without making too many assumptions about the acoustic properties of the instrument.

What do you think?

Reply:
Hi, I totally understand your point, but I opted to taking the running maximum for three reasons.

1. This approach seemed faster.  When generating piano rolls, this approach seems faster than keeping track of the current note on events, especially when the piano roll w/o pedal messages are already given.  
2. This approach covers the majority of use-case of CC64 on musical instruments. Using the most recent velo is a good assumption for decaying sounds that are played with a long interval.  Considering that pedals are often depressed for only a few bars, and also since CC64 are also used extensively on non-decaying instruments like string pads and organs, I argue that taking the running maximum captures the majority of the use-cases of CC64.    
3. I think the two approaches are complementary ways of approximating the pedals, but the choice would not influence most use-cases of how piano-rolls are used, as binary indicator of note presence.  If knowing the actual loudness is important, then one should convolve the note-on messages with the ADSR envelope.  But this kind of implementation is too time-consuming to be valuable in most cases of interest of using the piano-rolls like AMT or informed source separation; the interest is in knowing the existence of notes and not the loudness.

Reply:
Yup, I can see the arguments that way too. I'll leave it to @craffel to decide. :-)